### DISCUSSION TITLE
Issue compiling on CLI

### DISCUSSION THREAD

**Discussion by epudwjuhc** [arendst/Tasmota#13601] [Q&A] [answered]
Hi

tried to build Tasmonta just on CLI, using the  platformio for VIM setup (following the instructions! https://platformio.org/install/integration ).

It is just not working, something is missing in the config for the build system:

`
Compiling .pio/build/sonoff_basic/src/lib/default/Ext-printf/src/ext_printf.cpp.o
Compiling .pio/build/sonoff_basic/src/lib/default/Ext-printf/test/test_ext_printf.cpp.o
src/lib/default/Ext-printf/test/test_ext_printf.cpp:20:10: fatal error: ext_printf.h: No such file or directory

********************************************************************
* Looking for ext_printf.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:ext_printf.h"
* Web  > https://platformio.org/lib/search?query=header:ext_printf.h
*
********************************************************************

   20 | #include "ext_printf.h"
      |          ^~~~~~~~~~~~~~
`

and so on, for tons of header files... some of them can be fixed by putting the #include in "" instead of <>, or vice versa...

now please do not tell me I have to install Eclipse or Vscode or whatever to just generate firmware for a 1MB flash device... (u must be kidding me!).

Any idea why it is not finding header files (and most of them are just in the same directory as the source being compiled so #include "xxxx.h" SHOULD find it!).

**Comment by ascillato** [comment, marked as answer]
>now please do not tell me I have to install Eclipse or Vscode or whatever to just generate firmware for a 1MB flash device... (u must be kidding me!).

If you want to flash a device, there is no need to compile. That is already provided. Also you can use Tasmotizer to download and flash latest precompiled firmwares or webinstaller (https://tasmota.github.io/install/). For more information on flashing Tasmota, please check the docs at https://tasmota.github.io/docs/Getting-Started/

If you want to compile by yourself but you don't want to install anything, you can use online VSCode or Gitpod or Tasmocompiler that all are web based. Please, check Tasmota docs for more information on compiling https://tasmota.github.io/docs/Compile-your-build/

**Comment by barbudor** [comment]
Considering that Tasmota can be build on Linux from the platformIO command line without any integration (such as docker-tasmota), I trust nothing is missing in the project.
The miss is probably somewhere on your side. As no one in the team is using vim, you are on your own here unfortunately.

**Reply by epudwjuhc** [reply]
pip install -U idna
pip install -U platformio
export PATH="/home/user/.local/bin:$PATH"

mkdir test
cd test
pio project init --ide vim --board sonoff_basic
pico Makefile

Makefile added as by https://docs.platformio.org/en/latest//integration/ide/vim.html

tasmota src dumped into test/src

note: vim integration is basically no integration, just adds a simple makefile.

pio is working however will not build the project

**Comment by epudwjuhc** [comment]
it should be working out of the box but it doesn't
`
pio run
Processing sonoff_basic (platform: espressif8266; board: sonoff_basic; framework: arduino)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/sonoff_basic.html
PLATFORM: Espressif 8266 (3.2.0) > Sonoff Basic
HARDWARE: ESP8266 80MHz, 80KB RAM, 1MB Flash
PACKAGES: 
 - framework-arduinoespressif8266 3.30002.0 (3.0.2) 
 - tool-esptool 1.413.0 (4.13) 
 - tool-esptoolpy 1.30000.201119 (3.0.0) 
 - toolchain-xtensa 2.100300.210717 (10.3.0)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 35 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <Wire> 1.0
|-- <EEPROM> 1.0
|-- <ESP8266WebServer> 1.0
|   |-- <ESP8266WiFi> 1.0
|-- <ESP8266WiFi> 1.0
|-- <ESP8266HTTPClient> 1.2
|   |-- <ESP8266WiFi> 1.0
|-- <LittleFS> 0.1.0
|-- <SD> 2.0.0
|   |-- <SDFS> 0.1.0
|   |   |-- <SPI> 1.0
|   |   |-- <ESP8266SdFat> 2.0.2
|   |   |   |-- <SPI> 1.0
|-- <SPI> 1.0
|-- <ArduinoOTA> 1.0
|   |-- <ESP8266WiFi> 1.0
|   |-- <ESP8266mDNS> 1.2
|   |   |-- <ESP8266WiFi> 1.0
|-- <Ticker> 1.0
Building in release mode
Compiling .pio/build/sonoff_basic/src/lib/default/Ext-printf/src/ext_printf.cpp.o
Compiling .pio/build/sonoff_basic/src/lib/default/Ext-printf/test/test_ext_printf.cpp.o
Compiling .pio/build/sonoff_basic/src/lib/default/TasmotaSerial-3.3.0/src/TasmotaSerial.cpp.o
src/lib/default/Ext-printf/test/test_ext_printf.cpp:20:10: fatal error: ext_printf.h: No such file or directory

********************************************************************
* Looking for ext_printf.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:ext_printf.h"
* Web  > https://platformio.org/lib/search?query=header:ext_printf.h
*
********************************************************************

   20 | #include "ext_printf.h"
      |          ^~~~~~~~~~~~~~
compilation terminated.
*** [.pio/build/sonoff_basic/src/lib/default/Ext-printf/test/test_ext_printf.cpp.o] Error 1
Compiling .pio/build/sonoff_basic/src/lib/default/base64-1.1.1/catch.cpp.o
src/lib/default/Ext-printf/src/ext_printf.cpp:23:10: fatal error: SBuffer.hpp: No such file or directory

*******************************************************************
* Looking for SBuffer.hpp dependency? Check our library registry!
*
* CLI  > platformio lib search "header:SBuffer.hpp"
* Web  > https://platformio.org/lib/search?query=header:SBuffer.hpp
*
*******************************************************************

   23 | #include <SBuffer.hpp>
      |          ^~~~~~~~~~~~~
compilation terminated.
*** [.pio/build/sonoff_basic/src/lib/default/Ext-printf/src/ext_printf.cpp.o] Error 1
src/lib/default/TasmotaSerial-3.3.0/src/TasmotaSerial.cpp:28:10: fatal error: TasmotaSerial.h: No such file or directory

***********************************************************************
* Looking for TasmotaSerial.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:TasmotaSerial.h"
* Web  > https://platformio.org/lib/search?query=header:TasmotaSerial.h
*
***********************************************************************

   28 | #include <TasmotaSerial.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pio/build/sonoff_basic/src/lib/default/TasmotaSerial-3.3.0/src/TasmotaSerial.cpp.o] Error 1
src/lib/default/base64-1.1.1/catch.cpp:3:10: fatal error: base64.hpp: No such file or directory

******************************************************************
* Looking for base64.hpp dependency? Check our library registry!
*
* CLI  > platformio lib search "header:base64.hpp"
* Web  > https://platformio.org/lib/search?query=header:base64.hpp
*
******************************************************************

    3 | #include "base64.hpp"
      |          ^~~~~~~~~~~~
compilation terminated.
`

**Comment by epudwjuhc** [comment]
what works a bit better is:

clone git repo then run

pio project init --board sonoff_basic

=> here already first bug, it will NOT set the board to sonoff_basic in platformio.ini! Instead it will put generic ESP8266 there! I had to edit it by hand!

then

pio run

this will start to compile but then bail out with the following message:

Compiling .pio/build/tasmota/libd1e/bearssl-esp8266/x509/x509_minimal.c.o
Compiling .pio/build/tasmota/libd1e/bearssl-esp8266/x509/x509_minimal_full.c.o
Compiling .pio/build/tasmota/lib53d/tls_mini/StackThunk_light.cpp.o
Compiling .pio/build/tasmota/lib53d/tls_mini/WiFiClientSecureLightBearSSL.cpp.o
Compiling .pio/build/tasmota/libd49/Wire/Wire.cpp.o
Archiving .pio/build/tasmota/lib53d/libtls_mini.a
Indexing .pio/build/tasmota/lib53d/libtls_mini.a
Compiling .pio/build/tasmota/lib4b7/Adafruit_CCS811-1.0.0.14/Adafruit_CCS811.cpp.o
/home/user/.platformio/packages/framework-arduinoespressif8266@2.7.4+9/libraries/Wire/Wire.cpp:38:2: error: #error Wire library is not supported on this board
 #error Wire library is not supported on this board
  ^

**Comment by epudwjuhc** [comment]
how do I disable the wire library?

**Reply by barbudor** [reply]
What are you trying to build?
Command should be 'platformio run -e tasmota'
Have you read the 'compile-your-buod' doc?

**Comment by epudwjuhc** [comment]
it build whatever is in the override config, tasmota or tasmota-lite - anyway same outcome. will just NOT compile for sonoff and fail on wire library

**Comment by epudwjuhc** [comment]
that's basically the platformio.ini that I get generated by pio:

; PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html

[platformio]
description = Provide ESP8266 / ESP32 based devices with Web, MQTT and OTA firmware
src_dir = tasmota
lib_dir = lib/default
boards_dir = boards
build_cache_dir = .cache
extra_configs = platformio_tasmota32.ini
	platformio_tasmota_env.ini
	platformio_tasmota_env32.ini
	platformio_override.ini

[common]
platform = ${core.platform}
platform_packages = ${core.platform_packages}
framework = arduino
board = esp8266_1M
board_build.filesystem = littlefs
custom_unpack_dir = unpacked_littlefs
build_unflags = ${core.build_unflags}
build_flags = ${core.build_flags}
monitor_speed = 115200
monitor_port = COM5
upload_resetmethod = nodemcu
upload_port = COM5
extra_scripts = ${scripts_defaults.extra_scripts}
lib_ldf_mode = chain
lib_compat_mode = strict
shared_libdeps_dir = lib
lib_extra_dirs = 
	lib/lib_basic
	lib/lib_i2c
	lib/lib_display
	lib/lib_ssl
	lib/lib_audio
	lib/lib_rf
	lib/lib_div

[scripts_defaults]
extra_scripts = pio-tools/strip-floats.py
	pio-tools/name-firmware.py
	pio-tools/gzip-firmware.py
	pio-tools/override_copy.py
	pio-tools/download_fs.py

[esp_defaults]
build_unflags = -Wall
	-Wdeprecated-declarations
build_flags = -DCORE_DEBUG_LEVEL=0
	-Wl,-Map,firmware.map
	-Wno-deprecated-declarations
	-mtarget-align
	-free
	-fipa-pta
	-Wreturn-type
	-DUSE_CONFIG_OVERRIDE

[esp82xx_defaults]
build_flags = ${esp_defaults.build_flags}
	-DNDEBUG
	-DFP_IN_IROM
	-DBEARSSL_SSL_BASIC
	-DPIO_FRAMEWORK_ARDUINO_ESPRESSIF_SDK22x_190703
	-DPIO_FRAMEWORK_ARDUINO_LWIP2_HIGHER_BANDWIDTH_LOW_FLASH
	-DVTABLES_IN_FLASH
	-DPSTR_ALIGN=1
	-DMIMETYPE_MINIMAL

[core]
platform = espressif8266 @ 2.6.3
platform_packages = tasmota/framework-arduinoespressif8266 @ ~2.7.4
build_unflags = ${esp_defaults.build_unflags}
build_flags = ${esp82xx_defaults.build_flags}
	-DWAVEFORM_LOCKED_PWM

[env:sonoff_basic]
platform = espressif8266
board = sonoff_basic
framework = arduino

**Comment by epudwjuhc** [comment]
if I run pio run -e tasmota-lite, I get a firmware binary:

l build_output/firmware/
total 1920
drwxr-xr-x 2 user users    120 Nov  8 12:06 ./
drwxr-xr-x 4 user users     80 Nov  8 09:16 ../
-rw-r--r-- 1 user users 512944 Nov  8 12:06 tasmota-lite.bin
-rw-r--r-- 1 user users 365979 Nov  8 12:06 tasmota-lite.bin.gz

however, this firmware is not flashable to sonoff mini - it will complain with:

Upload Failed
Not enough space

when I try to OTA from the Tasmota firmware upgrade button (the mini is already running stock tasmota but I want to try my own firmware!).

**Comment by epudwjuhc** [comment]
I guess this is due to

board = esp8266_1M

and not sonoff_basic 

(however it will NOT compile for sonoff basic and bail out on that wire.h stuff)

**Reply by barbudor** [reply]
Stop shouting and read the docs
Sonoff_basic does not exist anymore
The firmware for Sonoff Basic device is tasmota.bin

**Reply by epudwjuhc** [reply]
and the one I get from tasmota-master build doesn't flash - not enough space LOL

**Comment by barbudor** [comment]
If you want a binary that still fit in size with crappy sonoff DIY, use Tasmota-lite.bin from v9.1
Otherwise serial flash

**Reply by epudwjuhc** [reply]
it is already running tasmota 10.0.0, I Just want to replace stock tasmota with my custom tasmota. the full stock tasmota 10.0.0 fit the sonoff mini flash size. Why it would not fit the master-lite branch?

**Comment by epudwjuhc** [comment]
it flashed the mini from ITEAD to tasmota from these binaries:

http://ota.tasmota.com/tasmota/release/

first I flashed the lite, then from lite the full tasmota it is 

613k | 10.0.0 | 20211019 11:03

in size.

the tasmota buid from master is 627248 bytes in size, thats 613k too

**Reply by barbudor** [reply]
So if you already flashed a Tasmota on it you can always upgrade by ota
Always use the gz compressed binaries
For large firmware there is a 2 step upgrade with tasmota-minimal.bin.gz first, then your final firmware

**Reply by barbudor** [reply]
If you upgrade from web server, tasmota automatically takes care of the 2 steps for you
If you use file upload, you need to do it manually

**Comment by epudwjuhc** [comment]
ok, so you say the mistake is in first place trying to build for sonoff_mini, I should just use generic build for ESP8266. Then use the .gz firmware image to flash, not the uncompressed one. let me try that

**Reply by barbudor** [reply]
Yep

**Reply by barbudor** [reply]
There is no 'build for Sonoff mini'
Sonoff_basic is an old name
All fw are named by 'tasmota-xx' for esp8266 (with predefined variants such as lite, display, sensors.... With different sets of built-in features)
'tasmota32-xx' for esp32
'tasmota32s2-xxx' 'tasmota32c3-xxx' etc for new cpus

**Comment by epudwjuhc** [comment]
seems that works now, thanks
